classes/migration/upgrade/PKPv3_3_0UpgradeMigration.inc.php

```diff
@@ -168,4 +188,235 @@ private function _makeRemoteUrlLocalizable() {
+				->update(['setting_value' => $newFileId]);
+		}
+
+		// Collect rows that will be deleted because they are old revisions
+		// They are identified by the new_file_id column, which is the only unique
+		// column on the table at this point.
+		$newFileIdsToDelete = [];
+
+		// Get all the unique file_ids. For each one, determine the latest revision
+		// in order to keep it in the table. The others will be flagged for removal
+		$revisionRowFileIds = Capsule::table('submission_files')
+			->groupBy('file_id')
+			->pluck('file_id');
+		foreach ($revisionRowFileIds as $revisionRowFileId) {
+			$submissionFileRows = Capsule::table('submission_files')
+				->where('file_id', '=', $revisionRowFileId)
+				->orderBy('revision', 'desc')
+				->get([
+					'file_id',
+					'new_file_id',
+				]);
+			$latestFileId = $submissionFileRows[0]->new_file_id;
+			foreach ($submissionFileRows as $submissionFileRow) {
+				if ($submissionFileRow->new_file_id !== $latestFileId) {
+					$newFileIdsToDelete[] = $submissionFileRow->new_file_id;
+				}
+			}
+		}
+
+		// Delete the rows for old revisions
+		Capsule::table('submission_files')
+			->whereIn('new_file_id', $newFileIdsToDelete)
+			->delete();
+
+		// Set assoc_type and assoc_id for all review round files
+		// Run this before migration to internal review file stages
+		$rows = Capsule::table('review_round_files')->get();
+		foreach ($rows as $row) {
+			Capsule::table('submission_files')
+				->where('file_id', '=', $row->file_id)
+				->whereIn('file_stage', [SUBMISSION_FILE_REVIEW_FILE, SUBMISSION_FILE_REVIEW_REVISION])
+				->update([
+					'assoc_type' => ASSOC_TYPE_REVIEW_ROUND,
+					'assoc_id' => $row->review_round_id,
+				]);
+		}
+
+		// Update file stage for all internal review files
+		Capsule::table('submission_files as sf')
+			->leftJoin('review_round_files as rrf', 'sf.file_id', '=', 'rrf.file_id')
+			->where('sf.file_stage', '=', SUBMISSION_FILE_REVIEW_FILE)
+			->where('rrf.stage_id', '=', WORKFLOW_STAGE_ID_INTERNAL_REVIEW)
+			->update(['sf.file_stage' => SUBMISSION_FILE_INTERNAL_REVIEW_FILE]);
+		Capsule::table('submission_files as sf')
+			->leftJoin('review_round_files as rrf', 'sf.file_id', '=', 'rrf.file_id')
+			->where('sf.file_stage', '=', SUBMISSION_FILE_REVIEW_REVISION)
+			->where('rrf.stage_id', '=', WORKFLOW_STAGE_ID_INTERNAL_REVIEW)
+			->update(['sf.file_stage' => SUBMISSION_FILE_INTERNAL_REVIEW_REVISION]);
+
+		// Update name of event log params to reflect new file structure
+		Capsule::table('event_log_settings')
+			->where('setting_name', 'fileId')
+			->update(['setting_name' => 'submissionFileId']);
+		Capsule::table('event_log_settings')
+			->where('setting_name', 'fileRevision')
+			->update(['setting_name' => 'fileId']);
+
+		// Restructure submission_files and submission_file_settings tables
+		Capsule::schema()->table('submission_files', function (Blueprint $table) {
+			$table->renameColumn('file_id', 'submission_file_id');
```

Comment:
@NateWr, have you tried running the migrations with PostgreSQL? I suspect it might get unhappy about some of the renames of primary keys for the sake of the implicit sequences and indexes it creates to go with them.

Reply:
@asmecher I'm working on the `postgres` migration now and hitting a blocker I can't figure out. The DAO method `getInsertId()` is throwing an error:

```
PHP Warning:  pg_query(): Query failed: ERROR:  currval of sequence &quot;submission_files_file_id_seq&quot; is not yet defined in this session in /home/nate/Projects/pkp/ojs/lib/pkp/lib/vendor/adodb/adodb-php/drivers/adodb-postgres7.inc.php on line 259
```

And this leads to a fatal error when it tries to insert into the settings table without an id. The interesting thing is that the sequence it is trying to get, `submission_files_file_id_seq`, no longer exists. During the migration, the column sequence is updated to `submission_files_submission_file_id_seq` (`file_id` => `submission_file_id`).

Here's the migration code I'm using:

```
		// Restructure submission_files and submission_file_settings tables
		Capsule::schema()->table('submission_files', function (Blueprint $table) {
			$table->renameColumn('file_id', 'submission_file_id');
			$table->renameColumn('new_file_id', 'file_id');
			$table->bigInteger('file_id')->nullable(false)->change();
			$table->renameColumn('source_file_id', 'source_submission_file_id');
			$table->renameColumn('date_uploaded', 'created_at');
			$table->renameColumn('date_modified', 'updated_at');
			$table->dropColumn('revision');
			$table->dropColumn('source_revision');
			$table->dropColumn('file_size');
			$table->dropColumn('file_type');
			$table->dropColumn('original_file_name');
			$table->foreign('file_id')->references('file_id')->on('files');
		});
		// Add primary key column attributes in separate migration to
		// avoid error in postgres with unfound column
		Capsule::schema()->table('submission_files', function (Blueprint $table) {
			$table->increments('submission_file_id')->change();
			$table->primary('submission_file_id');
			$table->unique('submission_file_id');
		});
```

For some reason, when renaming `file_id` to `submission_file_id`, the auto-increment is lost (this only seems to happen in postgres). That's why I've re-added it with `$table->increments('submission_file_id')`. And you can see that the table has the new sequence name:

```
                                                        Table "public.submission_files"
          Column           |            Type             | Collation | Nullable |                           Default                            
---------------------------+-----------------------------+-----------+----------+--------------------------------------------------------------
 submission_file_id        | integer                     |           | not null | nextval('submission_files_submission_file_id_seq'::regclass)
 source_submission_file_id | bigint                      |           |          | 
 submission_id             | bigint                      |           | not null | 
 genre_id                  | bigint                      |           |          | 
 file_stage                | bigint                      |           | not null | 
 direct_sales_price        | character varying(255)      |           |          | 
 sales_type                | character varying(255)      |           |          | 
 viewable                  | smallint                    |           |          | 
 created_at                | timestamp without time zone |           | not null | 
 updated_at                | timestamp without time zone |           | not null | 
 uploader_user_id          | bigint                      |           |          | 
 assoc_type                | bigint                      |           |          | 
 assoc_id                  | bigint                      |           |          | 
 file_id                   | bigint                      |           |          | 
Indexes:
    "submission_files_pkey" PRIMARY KEY, btree (submission_file_id)
    "submission_files_submission_file_id_unique" UNIQUE CONSTRAINT, btree (submission_file_id)
    "submission_files_stage_assoc" btree (file_stage, assoc_type, assoc_id)
    "submission_files_submission_id" btree (submission_id)
Foreign-key constraints:
    "submission_files_file_id_foreign" FOREIGN KEY (file_id) REFERENCES files(file_id)
Referenced by:
    TABLE "publication_galleys" CONSTRAINT "publication_galleys_submission_file_id_foreign" FOREIGN KEY (submission_file_id) REFERENCES submission_files(submission_file_id)
    TABLE "review_files" CONSTRAINT "review_files_submission_file_id_foreign" FOREIGN KEY (submission_file_id) REFERENCES submission_files(submission_file_id)
    TABLE "review_round_files" CONSTRAINT "review_round_files_submission_file_id_foreign" FOREIGN KEY (submission_file_id) REFERENCES submission_files(submission_file_id)
    TABLE "submission_file_revisions" CONSTRAINT "submission_file_revisions_submission_file_id_foreign" FOREIGN KEY (submission_file_id) REFERENCES submission_files(submission_file_id)
```

Any idea how to address this?

Reply:
I forgot to add that the sequence appears to work alright. The new submission file is given the correct `submission_file_id` when added. It's just that `getInsertId()` seems to fail because it is looking for the wrong sequence name (I guess).

Reply:
This is due to PostgreSQL's autoincrement handling -- it creates a sequence and index for every autoincrement, then sets up a trigger that automatically draws a new value from the sequence when inserting a fresh row. But when things get renamed incompletely (i.e. the sequence falls out of sync with the table or trigger) this happens.

In the past we've done something like the following to allow PostgreSQL to continue naming things as it likes (without having to hard-code anything relevant):
- Rename `submission_files` to `submission_files_tmp`
- Create a new `submission_files` table with the new schema
- Use `SELECT INTO` to migrate data from the old schema into the new schema
- For PostgreSQL users only, reset the sequence so the next insert will get the right number. This can be done without naming the sequence explicitly e.g. here: https://github.com/pkp/ojs/blob/master/dbscripts/xml/upgrade/3.0.0_preupdate.xml#L82

Reply:
Thanks Alec! That was the hint I needed. In the end, I decided to try solving this without rebuilding the whole table, since the new table has the correct sequence in it. So I [dropped the old sequence manually](https://github.com/pkp/pkp-lib/pull/6292/commits/93f6725c39ff7b75dabbee9e35186b61f6c4e434#diff-f4951b8d90b205bf60a1111a761b62eb75f598c0c536738da01d027561a33f1cR423). I still had an error when `DAO::getInsertId()` was not returning an ID, so I used the [pdo object from Laravel directly](https://github.com/pkp/pkp-lib/pull/6292/commits/93f6725c39ff7b75dabbee9e35186b61f6c4e434#diff-fa39ec93e3ba4e65f4455653e5700e9ac8a43cb09a3abc72040b0f31dc8628beR79-R80). This seemed to get the job done.

Any objections to that solution? It's only applied on DAOs that extend the `SchemaDAO`.

I've run successful migrations and installs for mysql and postgres now. Probably a good idea for you to look over [all the migrations changes I made working through this](https://github.com/pkp/pkp-lib/pull/6292/commits/93f6725c39ff7b75dabbee9e35186b61f6c4e434).

Reply:
Hmm, a missing ID from `DAO::getInsertId()` probably indicates that there is still something amiss in the sequence/trigger configuration... [Here is the ADODB code](https://github.com/ADOdb/ADOdb/blob/master/drivers/adodb-postgres64.inc.php#L168..L178) that backs `DAO::getInsertId()`, which hard-codes the sequence name.

I would suggest taking 1) a fresh install schema, and 2) a migrated schema, and comparing the two against each other to see how they're different. In PostgreSQL you can do this e.g. with:
```
\d submission_files
```
This will show you the trigger under the `Default` column for the ID row:
```
file_id            | bigint                         |           | not null | nextval('submission_files_file_id_seq'::regclass)
```
From there, you can check to see that the sequence is still named as expected after the migration runs.

Reply:
I had checked that. You can see that after the migration the sequence is correctly named `submission_files_submission_file_id_seq`:

```
                                                        Table "public.submission_files"
          Column           |            Type             | Collation | Nullable |                           Default                            
---------------------------+-----------------------------+-----------+----------+--------------------------------------------------------------
 submission_file_id        | integer                     |           | not null | nextval('submission_files_submission_file_id_seq'::regclass)
```

And yet the `getInsertId()` method was returning an error that mentioned the old sequence: `submission_files_file_id_seq`. Dropping that sequence during migration stopped the error message from appearing, but `getInsertId()` was still not returning anything.

I can provide you with a postgres database ready to be migrated if you'd like to try it out and see? I'm running postgres 12.4 if that matters.

Reply:
Oh wait, I'm not seeing the sequence in my latest upgraded database. Ugh... let me check.

Reply:
Ok, so I've gone in and restored the sequence. So now I have the updated sequence in my table:

```
                                                        Table "public.submission_files"
          Column           |            Type             | Collation | Nullable |                           Default                            
---------------------------+-----------------------------+-----------+----------+--------------------------------------------------------------
 submission_file_id        | bigint                      |           | not null | nextval('submission_files_submission_file_id_seq'::regclass)
```

However, `getInsertId()` is still returning `null`. I've checked the `tableName` and `primaryKeyColumn` values that are passed down to ADODB, and I tried running the ADODB query manually and the query returns the correct value:

```
ojs_file_upgrade=# select last_value from submission_files_submission_file_id_seq;
 last_value 
------------
          7
(1 row)
```

Pretty baffled here. I pushed up the fix to restore the sequence. Can I pass you a database? Here's the diff that reverts `SchemaDAO` to use `getInsertId()`:

```diff
diff --git a/classes/db/SchemaDAO.inc.php b/classes/db/SchemaDAO.inc.php
index 76e496d4c..18b1c447d 100644
--- a/classes/db/SchemaDAO.inc.php
+++ b/classes/db/SchemaDAO.inc.php
@@ -77,7 +77,8 @@ abstract class SchemaDAO extends DAO {
                }
 
                Capsule::table($this->tableName)->insert($primaryDbProps);
-               $object->setId(Capsule::getPdo()->lastInsertId());
+               $object->setId($this->getInsertId());
+               // $object->setId(Capsule::getPdo()->lastInsertId());
 
                // Add additional properties to settings table if they exist
                if (count($sanitizedProps) !== count($primaryDbProps)) {
@@ -264,7 +265,8 @@ abstract class SchemaDAO extends DAO {
         * @return int
         */
        public function getInsertId() {
-               return Capsule::getPdo()->lastInsertId();
+               return $this->_getInsertId($this->tableName, $this->primaryKeyColumn);
+               // return Capsule::getPdo()->lastInsertId();
        }
 
        /**
```

Reply:
Yes, please send me a DB dump!

Reply:
Thanks! I've sent that via db dump as well as instructions on making sure all the right branches are checked out. :)